File: examples/ocaml/linear/sparse.ml

open Sundials
open Bigarray

module S = Matrix.Sparse
module D = Matrix.Dense
let printf = Format.printf
let fprintf = Format.fprintf

let print_mat out mat =
  let m, n = S.size mat in
  let nnz, _ = S.dims mat in
  fprintf out "@[<v>matrix (M=%d, N=%d, NNZ=%d):@\n" m n nnz;
  for i = 0 to n - 1 do
    fprintf out "  col %d:" i;
    for j = S.get_col mat i to (S.get_col mat (i + 1)) - 1 do
      let r, v = S.get mat j in
      printf " (%d: % .02e)" r v
    done;
    fprintf out "@\n"
  done;
  fprintf out "@]"

let print_vec out m =
  let nr = Array1.dim m in
  fprintf out "@[<h>";
  for i = 0 to nr - 1 do
    fprintf out "@ % e" m.{i}
  done;
  fprintf out "@]"

let nrows, ncols, nzeros = 3, 3, 5;;

let main () =
  let da = D.create nrows ncols in
  let b = S.(make CSC nrows ncols 2) in
  let zero = S.(make CSC nrows ncols 0) in
  let x = Sundials.RealArray.of_list [2.0; 3.0; 4.0] in
  let y = Sundials.RealArray.create nrows in

  D.set da 0 0 ( 0.0);
  D.set da 0 1 ( 2.0);
  D.set da 0 2 ( 0.0);

  D.set da 1 0 ( 0.0);
  D.set da 1 1 (-4.0);
  D.set da 1 2 ( 0.0);

  D.set da 2 0 ( 0.0);
  D.set da 2 1 (-9.0);
  D.set da 2 2 (-3.0);

  printf "initially da=@\n%a@\n%!" D.pp da;

  let a = S.(from_dense CSC 0.0 da) in
  printf "initially a=@\n%a@\n%a@\n%!" S.pp a print_mat a;

  S.scale_addi 1.0 a;
  printf "a + 1=@\n%a@\n" print_mat a;

  S.blit ~src:a ~dst:b;
  S.scale_add 2.0 b zero;
  printf "scale copy x2: b=@\n%a@\n" print_mat b;

  S.set_to_zero b;
  S.resize ~nnz:9 b;
  printf "set to zero (NNZ=9): b=@\n%a@\n" print_mat b;

  S.set_col b 0 0;
  S.set_col b 1 0;
  S.set_col b 2 0;
  S.set_col b 3 1;

  S.set b 0 1 7.0;

  printf "b with element [r=1, c=2] set to 7.0:@\n%a@\n" print_mat b;

  S.scale_add 1.0 a b;
  printf "a = a + b: a=@\n%a@\n" print_mat a;

  S.matvec a x y;
  printf "y = A*x: x=@\n%a@\ny=@\n%a@\n" print_vec x print_vec y;

  ();;

main ();;
Gc.compact ();;

File: examples/ocaml/linear/sparse.c
#include <stdio.h>
#include <stdlib.h>
#include <sundials/sundials_types.h>
#include <sundials/sundials_sparse.h>
#include <sundials/sundials_direct.h>
#include "../../../src/sundials/sundials_ml.h"

#define NROWS  3
#define NCOLS  3
#define NZEROS 5

realtype a_init[NROWS][NCOLS] = {
    {  0.0,  2.0,  0.0},
    {  0.0, -4.0,  0.0},
    {  0.0, -9.0, -3.0}
};

void print_vec(realtype* m, long int nr)
{
    int i;

    for (i=0; i < nr; ++i) {
	printf(" % e", m[i]);
    }
    printf("\n");
}

void print_pivots(long int* m, long int nr)
{
    int i;

    for (i=0; i < nr; ++i) {
	printf(" % ld", m[i]);
    }
    printf("\n");
}

void print_mat(SlsMat m)
{
    int i, j;

    printf("matrix (M=%d, N=%d, NNZ=%d):\n", m->M, m->N, m->NNZ);
    for (i = 0; i < m->N; ++i) {
	printf("  col %d:", i);
#if SUNDIALS_LIB_VERSION >= 270
	for (j = m->indexptrs[i]; j < m->indexptrs[i+1]; ++j) {
	    printf(" (%d: % .02e)", m->indexvals[j], m->data[j]);
	}
#else
	for (j = m->colptrs[i]; j < m->colptrs[i+1]; ++j) {
	    printf(" (%d: % .02e)", m->rowvals[j], m->data[j]);
	}
#endif
	printf("\n");
    }
}

int main(int argc, char** argv)
{
    DlsMat da = NewDenseMat(NROWS, NCOLS);
    SlsMat a;
#if SUNDIALS_LIB_VERSION >= 270
    SlsMat b  = SparseNewMat(NROWS, NCOLS, 2, CSC_MAT);
#else
    SlsMat b  = NewSparseMat(NROWS, NCOLS, 2);
#endif
    realtype x[NCOLS] = { 2.0, 3.0, 4.0 };
    realtype y[NROWS];
    int i, j;

    for (i=0; i < NROWS; ++i) {
	for (j=0; j < NCOLS; ++j) {
	    DENSE_ELEM(da, i, j) = a_init[i][j];
	}
    }

    printf("initially da=\n");
    PrintMat(da);
    printf("\n");

#if SUNDIALS_LIB_VERSION >= 270
    a = SparseFromDenseMat(da, CSC_MAT);
    printf("initially a=\n");
    SparsePrintMat(a, stdout);
#else
    a = SlsConvertDls(da);
    printf("initially a=\n");
    PrintSparseMat(a);
#endif
    print_mat(a);
    printf("\n");

#if SUNDIALS_LIB_VERSION >= 270
    SparseAddIdentityMat(a);
#else
    AddIdentitySparseMat(a);
#endif
    printf("a + 1=\n");
    print_mat(a);
    printf("\n");

#if SUNDIALS_LIB_VERSION >= 270
    SparseCopyMat(a, b);
    SparseScaleMat(2.0, b);
#else
    CopySparseMat(a, b);
    ScaleSparseMat(2.0, b);
#endif
    printf("scale copy x2: b=\n");
    print_mat(b);
    printf("\n");

#if SUNDIALS_LIB_VERSION >= 270
    SparseSetMatToZero(b);
#else
    SlsSetToZero(b);
#endif
    b->NNZ = 9;
#if SUNDIALS_LIB_VERSION >= 270
    b->indexvals = realloc(b->indexvals, b->NNZ*sizeof(int));
#else
    b->rowvals = realloc(b->rowvals, b->NNZ*sizeof(int));
#endif
    b->data    = realloc(b->data, b->NNZ*sizeof(realtype));
    printf("set to zero (NNZ=9): b=\n");
    print_mat(b);
    printf("\n");

#if SUNDIALS_LIB_VERSION >= 270
    b->indexptrs[0] = 0;
    b->indexptrs[1] = 0;
    b->indexptrs[2] = 0;
    b->indexptrs[3] = 1;

    b->indexvals[0] = 1;
#else
    b->colptrs[0] = 0;
    b->colptrs[1] = 0;
    b->colptrs[2] = 0;
    b->colptrs[3] = 1;

    b->rowvals[0] = 1;
#endif
    b->data[0] = 7.0;

    printf("b with element [r=1, c=2] set to 7.0:\n");
    print_mat(b);
    printf("\n");

#if SUNDIALS_LIB_VERSION >= 270
    SparseAddMat(a, b);
#else
    SlsAddMat(a, b);
#endif
    printf("a = a + b: a=\n");
    print_mat(a);
    printf("\n");

#if SUNDIALS_LIB_VERSION >= 270
    SparseMatvec(a, x, y);
#else
    SlsMatvec(a, x, y);
#endif
    printf("y = A*x: x=\n");
    print_vec(x, NCOLS);
    printf("y=\n");
    print_vec(y, NROWS);

    DestroyMat(da);
#if SUNDIALS_LIB_VERSION >= 270
    SparseDestroyMat(a);
    SparseDestroyMat(b);
#else
    DestroySparseMat(a);
    DestroySparseMat(b);
#endif

    return 0;
}

